Show unfinished items steps in todolist

MarcelBolten opened the pull request:
See #2804

I also did a small todolist UI overhaul.
<img src="https://user-images.githubusercontent.com/65481677/129155960-48257947-3556-4232-85e0-e426722f04cf.png" height="250">
@NicolasCARPi what do you think?

Todo:
- [ ]  Implementation of user option "Show team steps as default"

NicolasCARPi commented on src/models/Steps.php:
```diff
@@ -94,22 +98,57 @@ public function read(ContentParamsInterface $params): array
+     * @param bool $team Steps from the team or only user, only for items
      */
-    public function readAll(): array
+    public function readAll(string $table, bool $team = false): array
```
having a boolean flag in a function is a code smell and should be avoided. I realize it's easier said than done. But new code must not make the class rating go down by adding complexity:

![2021-08-12-110211_1110x486_scrot](https://user-images.githubusercontent.com/3043706/129169375-0e406638-3cd8-41f6-a553-e6385e58a19a.png)

So there should be a readAllFromTeam I think.

MarcelBolten replied:
Code is refactored.
That was a good exercise.
Maybe I overdid it but the addition of `getReadAllSql` and `cleanUpReadAllSQLResults` methods helps to avoid code duplication.

MarcelBolten commented on src/models/Steps.php:
```diff
@@ -98,79 +97,66 @@ public function read(ContentParamsInterface $params): array
+
+        $res = $req->fetchAll();
+        if ($res === false) {
+            return array();
```
Maybe this could go into `cleanUpReadAllSQLResults` too?

MarcelBolten commented on src/models/Steps.php:
```diff
@@ -98,79 +97,66 @@ public function read(ContentParamsInterface $params): array
         $res = $req->fetchAll();
         if ($res === false) {
             return array();
         }
```
see above

NicolasCARPi commented on src/templates/ucp.html:
```diff
@@ -198,6 +198,11 @@ <h3>{{ 'Miscellaneous'|trans }}</h3>
         />
         <label for='show_public'>{{ 'Show experiments/database items from outside the team'|trans }}</label>
         <br>
+        <input id='items_steps_scope' type='checkbox' name='items_steps_scope'
```
I don't like `items_steps_scope` as a name, because it is not obvious what it does. I suggest: `todolist_steps_show_team` instead, and why not extend this switch to include experiments and items at the same time?

MarcelBolten replied:
> why not extend this switch to include experiments and items at the same time

I'm not sure what you mean here.

NicolasCARPi replied:
The user <=> team switch could work for both experiments and items. So you can see only the unfinished steps for your experiments or for other experiments in the team/with read access. So one switch will control both experiments and items scope.

NicolasCARPi commented on src/models/Steps.php:
```diff
@@ -198,4 +223,48 @@ private function updateBody(string $content): bool
+     */
+    private function cleanUpReadAllSQLResults(array $res): array
+    {
+        foreach ($res as &$exp) {
```
exp should be entity now

NicolasCARPi commented on src/models/Steps.php:
```diff
@@ -198,4 +223,48 @@ private function updateBody(string $content): bool
+     * @param string $table Which table do we read from? experiments or items
+     * @param string $whereClause SQL WHERE clause
+     */
+    private function getReadAllSql(string $table, string $whereClause): string
```
this is ugly but I know very well that sometimes it is necessary to write ugly code, and I don't have a better suggestion at the moment. Anyway, I believe in iterative code so if it works like that it can be improved later. Maybe an implementation with classes would be cleaner.

NicolasCARPi commented on src/models/Steps.php:
```diff
@@ -90,41 +93,63 @@ public function read(ContentParamsInterface $params): array
+        $req->bindParam(':teamid', $this->Entity->Users->team, PDO::PARAM_INT);
+        $this->Db->execute($req);
+
+        $res = $req->fetchAll();
```
you can now use `$this->Db->fetchAll($req);` and remove the check for false. See 3cd60633dc690efd0124af67d7ef1d8e192364bb

NicolasCARPi commented on src/models/Steps.php:
```diff
@@ -77,8 +77,11 @@ public function import(array $step): void
         if ($params->getTarget() === 'all') {
-            return $this->readAll();
+            return $this->readAll($this->Entity->type);
+        } elseif ($params->getTarget() === 'all_team') {
```
else is not necessary here, you can close the if and make another if

NicolasCARPi reviewed (changes requested):
First I think it's a good addition to the codebase, and it's a feature I want to see in eLab because I use steps in database items for Deltablot.

But I think the way you implemented it can still be improved. I realize that it's easier said than done and that the code you're changing is not great either. But we should strive for SOLID code, with immutable, encapsulated objects properly using inheritance. I have found that it's easier to design a good class when you try to test it 100%. For instance I'm quite happy with https://github.com/elabftw/elabftw/blob/hypernext/src/services/ResetPasswordKey.php, it is testable, doesn't have side effects and the functions are pure. Also it is short and to the point. 

So what if we have UnfinishedSteps, which can take Experiments or Items as injected dependency, with a switch for the team scope, and a single public function to get them. It would be much cleaner.

MarcelBolten commented:
Hi @NicolasCARPi ,
I reworked the code:
UnfinishedSteps extends Steps now to keep the CRUD interface intact although we don't want to create, updated or destroy an unfinished step. But with only a readable interface the AbstractProcessor shows warnings/errors in the static analysis.
To pass the scope into the class I use the content entry of the payload to avoid extending the interface. This might be confusing though as content is usually used to provide content for create or update but not to define what content to get.
What do you prefer, extending the interface (target I think) or keeping it the way it is?

NicolasCARPi commented on src/models/UnfinishedSteps.php:
```diff
@@ -0,0 +1,90 @@
+
+        // Clean up the results so we get a nice array with entity id/title and steps with their id/body
+        // use reference to edit in place
+        foreach ($res as &$entity) {
```
I suggest to extract this in another function, maybe a private one: `format(array $steps: array`, and at the end of read you just return its result.

NicolasCARPi reviewed (changes requested):
Note: not a complete review

NicolasCARPi commented:
> UnfinishedSteps extends Steps now to keep the CRUD interface intact although we don't want to create, updated or destroy an unfinished step. But with only a readable interface the AbstractProcessor shows warnings/errors in the static analysis.

It's how it should be I think.

> To pass the scope into the class I use the content entry of the payload to avoid extending the interface. This might be confusing though as content is usually used to provide content for create or update but not to define what content to get.

I agree that it's confusing. I would use extraParams with `scope` => `team`.

MarcelBolten commented:
I tried to use extraParams but I can't make it work in a reasonable way. Some interface is always violated. Using just the read method does not allow to inject the extraParams.
I could make it work when injecting it straight into the new unfinishedSteps class but that defeats the purpose of heaving the getParams method.
One option is to create a new target or use readAll but that feels circumstantial.
After all unfinishedSteps should simply read the unfinished steps and have a simple switch for the scope.
@NicolasCARPi Do you have a another suggestion how to implement the switch?

NicolasCARPi commented:
It's fine to add a target then.

MarcelBolten commented:
After fiddling around some more I realized that the use of a target also did not work.
In the end I extended the ContentParamsInterface and allowed the ContentParams class to accepted extraParams as a parameter.
To me it seems to be the cleanest implementation.

MarcelBolten commented:
Should now work again. You can give it a try.
It would be nice to have a shortcut for the favorite tags panel.
And how does one deselect a favorite tag? Of course one can click on the experiment/database link in the nav bar but that feels odd. Maybe there can be a >clear selection< button?
